pkg/backend/openshiftcluster.go

```diff
@@ -235,6 +259,9 @@ func (ocb *openShiftClusterBackend) updateAsyncOperation(ctx context.Context, lo
 				asyncdoc.OpenShiftCluster = &ocCopy
 			}
 
+			clusterResult := newOperationResult(resultType, initialProvisioningState.String(), backendErr)
+			log.Printf("Long running result: %s", clusterResult.String())
```

Comment:
> Question: can you use error code to distinguish between the errors? CloudErrorCategory looks like a more generic duplicate of error code to me.

Nope. `CloudErrorCodes` lives in the repo and there is no easy way from kusto/metric side to extract that unless hardcoding them which is also no ideal. Yes, a bit of duplicate can get us so much benefit in terms of readability and usability.

> Is this about async operations specifically? E.g. you want to have more visibility into cluster creation/deletion/update success/failures?

yes, we want to track the result of cluster operations such as Create, Delete, Update, etc.

> Question: Can you use metrics for what you are doing?

Metric is only useful at high level view and alerting, but it loses the slice and dice capabilities from kusto to do further analysis

> If we absolutely have to use logs - then we need to change the approach. Instead of logging type of the result and operation name in the message we should look into adding new log fields (e.g log.WithField("...", "...")).

Ideally, we want this event in a different table. Typically, in Azure (CRP, AKS, NRP, you name it), there is a table like `AsyncQosEvents` that records the outcome of long running operations. We would also like to do that for ARO, but getting insights first will provide biggest bang of the bucks.
I don't get why it _has_ to be using `.WithFields()`. We already print the error in `MESSAGE` today, and we really care less of adding more columns to `ServiceLogs` table. Again, the goal is to write them into different table and we can either use Kusto to re-ingest or update mdsd/fluent-bit to do that, whichever is easier.

All we are doing is bringing all the experiences from Azure to ARO that will only help the service :)

Reply:
To learn more on what we want to achieve, you can also read more in the Resource Provider Level section in https://docs.google.com/document/d/12UG4rpU9uQCziJ0EKsU-rxSqIO1dPgaoo4-2yi1nzr4/edit

Reply:
> Nope. `CloudErrorCodes` lives in the repo and there is no easy way from kusto/metric side to extract that unless hardcoding them which is also no ideal. Yes, a bit of duplicate can get us so much benefit in terms of readability and usability.

Yes, `CloudErrorCodes` live in the ARO-RP repo, but this is an API: it is very unlikely that we will break support for this. If we do - it will be fairly high priority for us to fix. Besides that - to distinguish between "client errors" and everything else we don't need to know all the error codes. The logic is basically:

* If the error has `api.CloudErrorCodeInternalServerError` error code - it is an RP error
* If the error has the error code which is not `api.CloudErrorCodeInternalServerError` - it is a 'client error'"

I understand that we would like to avoid hardcoding. But I also want to avoid the need to keep error codes in sync with `CloudErrorCategory` across the codebase and having everyone on the team to remember that they need to make sure that their new error is being logged with the right category.

> Ideally, we want this event in a different table. Typically, in Azure (CRP, AKS, NRP, you name it), there is a table like `AsyncQosEvents` that records the outcome of long running operations.

Yep, that is fine. I don't know how to do that, but I'm sure we can find a way.

> I don't get why it _has_ to be using `.WithFields()`. We already print the error in `MESSAGE` today, and we really care less of adding more columns to `ServiceLogs` table.

We can print error in the message. But we should not print category or other additional information as part of the message. The reason is the same why we don't put resource ID, correlation ID and other data into message itself and use structured logs - usability. I don't want to have extra noise in the message and want them to be human readable. I want to be able to easily filter messages in Geneva. And I want to make sure that we don't have to worry about breaking format and breaking downstream consumers which rely on a message being formatted in a specific way.

I'm looking at this mostly from the SRE perspective and manual/semi-manual consumption, but I think having a separate field will also be beneficial for programmatic consumption (e.g. less chances of breaking apps by breaking the format).

> To learn more on what we want to achieve, you can also read more in the Resource Provider Level section in https://docs.google.com/document/d/12UG4rpU9uQCziJ0EKsU-rxSqIO1dPgaoo4-2yi1nzr4/edit

Thanks for sharing this. I'll take a look.